File: include/extras-kio/libkrecentdocument.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const krecentdocument_enums = enums;
const std = @import("std");

/// https://api.kde.org/krecentdocument.html
pub const krecentdocument = struct {
    /// New constructs a new KRecentDocument object.
    ///
    /// ``` other: QtC.KRecentDocument ```
    pub fn New(other: ?*anyopaque) QtC.KRecentDocument {
        return qtc.KRecentDocument_new(@ptrCast(other));
    }

    /// New2 constructs a new KRecentDocument object and invalidates the source KRecentDocument object.
    ///
    /// ``` other: QtC.KRecentDocument ```
    pub fn New2(other: ?*anyopaque) QtC.KRecentDocument {
        return qtc.KRecentDocument_new2(@ptrCast(other));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: QtC.KRecentDocument, other: QtC.KRecentDocument ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KRecentDocument_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: QtC.KRecentDocument, other: QtC.KRecentDocument ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.KRecentDocument_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#recentUrls)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn RecentUrls(allocator: std.mem.Allocator) []QtC.QUrl {
        const _arr: qtc.libqt_list = qtc.KRecentDocument_RecentUrls();
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QUrl, _arr.len) catch @panic("krecentdocument.RecentUrls: Memory allocation failed");
        const _data: [*]QtC.QUrl = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#add)
    ///
    /// ``` url: QtC.QUrl ```
    pub fn Add(url: ?*anyopaque) void {
        qtc.KRecentDocument_Add(@ptrCast(url));
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#add)
    ///
    /// ``` url: QtC.QUrl, groups: []krecentdocument_enums.RecentDocumentGroup ```
    pub fn Add2(url: ?*anyopaque, groups: []i32) void {
        const groups_list = qtc.libqt_list{
            .len = groups.len,
            .data = groups.ptr,
        };
        qtc.KRecentDocument_Add2(@ptrCast(url), groups_list);
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#add)
    ///
    /// ``` url: QtC.QUrl, desktopEntryName: []const u8 ```
    pub fn Add3(url: ?*anyopaque, desktopEntryName: []const u8) void {
        const desktopEntryName_str = qtc.libqt_string{
            .len = desktopEntryName.len,
            .data = desktopEntryName.ptr,
        };
        qtc.KRecentDocument_Add3(@ptrCast(url), desktopEntryName_str);
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#add)
    ///
    /// ``` url: QtC.QUrl, desktopEntryName: []const u8, groups: []krecentdocument_enums.RecentDocumentGroup ```
    pub fn Add4(url: ?*anyopaque, desktopEntryName: []const u8, groups: []i32) void {
        const desktopEntryName_str = qtc.libqt_string{
            .len = desktopEntryName.len,
            .data = desktopEntryName.ptr,
        };
        const groups_list = qtc.libqt_list{
            .len = groups.len,
            .data = groups.ptr,
        };
        qtc.KRecentDocument_Add4(@ptrCast(url), desktopEntryName_str, groups_list);
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#clearEntriesOldestEntries)
    ///
    /// ``` maxEntries: i32 ```
    pub fn ClearEntriesOldestEntries(maxEntries: i32) bool {
        return qtc.KRecentDocument_ClearEntriesOldestEntries(@intCast(maxEntries));
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#removeFile)
    ///
    /// ``` url: QtC.QUrl ```
    pub fn RemoveFile(url: ?*anyopaque) void {
        qtc.KRecentDocument_RemoveFile(@ptrCast(url));
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#removeApplication)
    ///
    /// ``` desktopEntryName: []const u8 ```
    pub fn RemoveApplication(desktopEntryName: []const u8) void {
        const desktopEntryName_str = qtc.libqt_string{
            .len = desktopEntryName.len,
            .data = desktopEntryName.ptr,
        };
        qtc.KRecentDocument_RemoveApplication(desktopEntryName_str);
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#removeBookmarksModifiedSince)
    ///
    /// ``` since: QtC.QDateTime ```
    pub fn RemoveBookmarksModifiedSince(since: ?*anyopaque) void {
        qtc.KRecentDocument_RemoveBookmarksModifiedSince(@ptrCast(since));
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#clear)
    ///
    ///
    pub fn Clear() void {
        qtc.KRecentDocument_Clear();
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#maximumItems)
    ///
    ///
    pub fn MaximumItems() i32 {
        return qtc.KRecentDocument_MaximumItems();
    }

    /// [Qt documentation](https://api.kde.org/krecentdocument.html#dtor.KRecentDocument)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KRecentDocument ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KRecentDocument_Delete(@ptrCast(self));
    }
};

/// https://api.kde.org/krecentdocument.html#types
pub const enums = struct {
    pub const RecentDocumentGroup = enum {
        pub const Development: i32 = 0;
        pub const Office: i32 = 1;
        pub const Database: i32 = 2;
        pub const Email: i32 = 3;
        pub const Presentation: i32 = 4;
        pub const Spreadsheet: i32 = 5;
        pub const WordProcessor: i32 = 6;
        pub const Graphics: i32 = 7;
        pub const TextEditor: i32 = 8;
        pub const Viewer: i32 = 9;
        pub const Archive: i32 = 10;
        pub const Multimedia: i32 = 11;
        pub const Audio: i32 = 12;
        pub const Video: i32 = 13;
        pub const Photo: i32 = 14;
        pub const Application: i32 = 15;
    };
};

File: src_original/extras-kio/libkrecentdocument.cpp
#include <KRecentDocument>
#include <QDateTime>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <krecentdocument.h>
#include "libkrecentdocument.h"
#include "libkrecentdocument.hxx"

KRecentDocument* KRecentDocument_new(const KRecentDocument* other) {
    return new KRecentDocument(*other);
}

KRecentDocument* KRecentDocument_new2(KRecentDocument* other) {
    return new KRecentDocument(std::move(*other));
}

void KRecentDocument_CopyAssign(KRecentDocument* self, KRecentDocument* other) {
    *self = *other;
}

void KRecentDocument_MoveAssign(KRecentDocument* self, KRecentDocument* other) {
    *self = std::move(*other);
}

libqt_list /* of QUrl* */ KRecentDocument_RecentUrls() {
    QList<QUrl> _ret = KRecentDocument::recentUrls();
    // Convert QList<> from C++ memory to manually-managed C memory
    QUrl** _arr = static_cast<QUrl**>(malloc(sizeof(QUrl*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = new QUrl(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void KRecentDocument_Add(const QUrl* url) {
    KRecentDocument::add(*url);
}

void KRecentDocument_Add2(const QUrl* url, libqt_list /* of int */ groups) {
    QList<KRecentDocument::RecentDocumentGroup> groups_QList;
    groups_QList.reserve(groups.len);
    int* groups_arr = static_cast<int*>(groups.data);
    for (size_t i = 0; i < groups.len; ++i) {
        groups_QList.push_back(static_cast<KRecentDocument::RecentDocumentGroup>(groups_arr[i]));
    }
    KRecentDocument::add(*url, groups_QList);
}

void KRecentDocument_Add3(const QUrl* url, const libqt_string desktopEntryName) {
    QString desktopEntryName_QString = QString::fromUtf8(desktopEntryName.data, desktopEntryName.len);
    KRecentDocument::add(*url, desktopEntryName_QString);
}

void KRecentDocument_Add4(const QUrl* url, const libqt_string desktopEntryName, libqt_list /* of int */ groups) {
    QString desktopEntryName_QString = QString::fromUtf8(desktopEntryName.data, desktopEntryName.len);
    QList<KRecentDocument::RecentDocumentGroup> groups_QList;
    groups_QList.reserve(groups.len);
    int* groups_arr = static_cast<int*>(groups.data);
    for (size_t i = 0; i < groups.len; ++i) {
        groups_QList.push_back(static_cast<KRecentDocument::RecentDocumentGroup>(groups_arr[i]));
    }
    KRecentDocument::add(*url, desktopEntryName_QString, groups_QList);
}

void KRecentDocument_RemoveFile(const QUrl* url) {
    KRecentDocument::removeFile(*url);
}

void KRecentDocument_RemoveApplication(const libqt_string desktopEntryName) {
    QString desktopEntryName_QString = QString::fromUtf8(desktopEntryName.data, desktopEntryName.len);
    KRecentDocument::removeApplication(desktopEntryName_QString);
}

void KRecentDocument_RemoveBookmarksModifiedSince(const QDateTime* since) {
    KRecentDocument::removeBookmarksModifiedSince(*since);
}

void KRecentDocument_Clear() {
    KRecentDocument::clear();
}

int KRecentDocument_MaximumItems() {
    return KRecentDocument::maximumItems();
}

void KRecentDocument_Delete(KRecentDocument* self) {
    delete self;
}
